pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java

```diff
@@ -806,6 +848,153 @@ private List<String> getServerTag(String serverName) {
+          throw new RuntimeException("Creation time is not found");
+        }
+        consumingSegmentsAge.put(s, (int) (now - creationTime) / 60_000);
+      }));
+    } catch (Exception e) {
+      return null;
+    }
+    return consumingSegmentsAge;
+  }
+
+  @VisibleForTesting
+  ConsumingSegmentInfoReader getConsumingSegmentInfoReader() {
+    return _consumingSegmentInfoReader;
+  }
+
+  /**
+   * Fetches the consuming segment info for the table and calculates the number of offsets to catch up for each
+   * consuming segment. consumingSegmentZKMetadata is a map from consuming segments to be moved to their ZK metadata.
+   * Returns a map from segment name to the number of offsets to catch up for that consuming
+   * segment. Return null if failed to obtain info for any consuming segment.
+   */
+  @Nullable
+  private Map<String, Integer> getConsumingSegmentsOffsetsToCatchUp(String tableNameWithType,
+      Map<String, SegmentZKMetadata> consumingSegmentZKMetadata) {
+    ConsumingSegmentInfoReader consumingSegmentInfoReader = getConsumingSegmentInfoReader();
+    if (consumingSegmentInfoReader == null) {
+      LOGGER.warn("ConsumingSegmentInfoReader is null, cannot calculate consuming segments info for table: {}",
+          tableNameWithType);
+      return null;
+    }
+    Map<String, Integer> segmentToOffsetsToCatchUp = new HashMap<>();
+    try {
+      ConsumingSegmentInfoReader.ConsumingSegmentsInfoMap consumingSegmentsInfoMap =
+          consumingSegmentInfoReader.getConsumingSegmentsInfo(tableNameWithType, 30_000);
+      for (Map.Entry<String, SegmentZKMetadata> entry : consumingSegmentZKMetadata.entrySet()) {
+        String segmentName = entry.getKey();
+        List<ConsumingSegmentInfoReader.ConsumingSegmentInfo> consumingSegmentInfoList =
+            consumingSegmentsInfoMap._segmentToConsumingInfoMap.getOrDefault(segmentName, null);
+        SegmentZKMetadata segmentZKMetadata = entry.getValue();
+        if (segmentZKMetadata == null) {
+          LOGGER.warn("Cannot find SegmentZKMetadata for segment: {} in table: {}", segmentName, tableNameWithType);
+          return null;
+        }
+        String startOffset = segmentZKMetadata.getStartOffset();
+        if (startOffset == null) {
+          LOGGER.warn("Start offset is null for segment: {} in table: {}", segmentName, tableNameWithType);
+          return null;
+        }
+        if (consumingSegmentInfoList == null || consumingSegmentInfoList.isEmpty()) {
+          LOGGER.warn("No available consuming segment info from any server. Segment: {} in table: {}", segmentName,
+              tableNameWithType);
+          return null;
+        }
+        // this value should be the same regardless of which server the consuming segment info is from, use the
+        // first in the list here
+        int offsetsToCatchUp =
+            consumingSegmentInfoList.get(0)._partitionOffsetInfo._latestUpstreamOffsetMap.values()
```

Comment:
looks like we use consumingSegmentInfoReader to ask servers for upstream's latest offsets. 
But under the hood, server uses this util to get offsets. So would it be more efficient to use this here to get the offsets from upstream directly?
```
_streamConsumerFactory = StreamConsumerFactoryProvider.create(_streamConfig);
...
_partitionMetadataProvider = _streamConsumerFactory.createPartitionMetadataProvider(
        _clientId, _streamPatitionGroupId);
...
_partitionMetadataProvider.fetchStreamPartitionOffset(OffsetCriteria.LARGEST_OFFSET_CRITERIA, timeout);
...
```
_streamConfig is from tableConfig, but may need to fake _clientId/groupId values